packages/adminrouter/extra/src/README.md

```diff
@@ -261,31 +261,190 @@ as-is/usually do not follow any of these patterns. It is possible to standardise
+* namespace and name: `foo/bar`, resource address: `/baz/picture.jpg`
+* namespace and name: `foo/bar/baz`, resource address: `/picture.jpg`
+* namespace and name: `foo/bar/baz/picture.jpg`, resource address: `/`
+
+So the algorithm iteratively tries to resolve each "candidate" namespace and
+name and assumes that first-match wins.
+
+Assuming that we deployed `foo/bar` NGINX task, we will see two iterations:
+* first one with `foo` candidate application ID, which is not present in the
+  cluster
+* second one with `foo/bar` candidate application ID, which results in the
+  request being routed to the correct backend. In the case when user is trying
+  to access a non-existent service, there are going to be four iterations,
+  each one unsuccessful. After the fourth one the code will give up.
+
+There is a limit on the number of iterations that AR makes. At the time of
+writing, it is set to `10`. This means that an application ID cannot nest more
+than ten times and that everything after the tenth segment is always treated as
+a resource component.
+
+#### Resolving individual components
+During each iteration described in previous paragraphs, AR tries to resolve the
+component using the data stored in the cache and if necessary - from MesosDNS:
+* first step is to:
+  * check if a root Marathon task has following labels set:
+    * `DCOS_SERVICE_NAME` label contents is equal to given application ID string
+    * `DCOS_SERVICE_SCHEME` label is set
+    * `DCOS_SERVICE_PORT_INDEX` label is set
+  * task is in the state `TASK_RUNNING`.
+  If all the conditions are met then the request is routed using root Marathon
+  data. Please check the [section below](#recommended-way-to-expose-services)
+  for details on the meaning of these labels.
+* if the name of the service couldn't be resolved using root Marathon data, AR
+  scans the list of Mesos frameworks in search for the one with either
+  framework ID (higher priority) or framework name(lower priority) equal to
+  application ID string. If found, the next steps depend on the value of
+  `webui_url` field:
+  * field contains an empty string - algorithm tries to resolve the application
+    ID using MesosDNS
+  * field is `null` - resolving process for given `namespace and name` ends,
+    MesosDNS step is NOT executed
+  * field contains a valid address - request is proxied to the address pointed
+    by the field contents
+* resolving via MesosDNS relies on issuing an HTTP SRV query to MesosDNS. For
+  `namespace and name` string equal to `foo/bar/baz`, the request that is going to
+  be made is:
+  ```
+  http://localhost:8123/baz.bar.foo._tcp.marathon.mesos
+  ```
+  In case when MesosDNS returns no results, AR assumes that there is no
+  task/framework with given application ID running on DC/OS.
+
+#### Stale cache/broken cache cases
+The `/service` endpoint internally uses AR cache, so an AR failing to refresh
+it has the same effects on it like any other endpoint. Please check the [Cache
```

Comment:
❓ I'm a bit confused as to what each "it" is. Ideally replace "it"s with what they are - I guess the first "it" means "the cache".